exporter/xray/segment.go

```diff
@@ -0,0 +1,319 @@
+// the same trace, in 16 hexadecimal digits.
+func makeID(spanID trace.SpanID) string {
+	if spanID == zeroSpanID {
+		return ""
+	}
+	return hex.EncodeToString(spanID[0:8])
+}
+
+// MakeTraceID – A unique identifier that connects all segments and subsegments
+// originating from a single client request.
+//  * A trace_id consists of three numbers separated by hyphens. For example,
+//    1-58406520-a006649127e371903a2de979. This includes:
+//  * The version number, that is, 1.
+//  * The time of the original request, in Unix epoch time, in 8 hexadecimal digits.
+//  * For example, 10:00AM December 2nd, 2016 PST in epoch time is 1480615200 seconds,
+//    or 58406520 in hexadecimal.
+//  * A 96-bit identifier for the trace, globally unique, in 24 hexadecimal digits.
+func MakeTraceID(traceID trace.TraceID, startTime time.Time) string {
+	const (
+		digits24 = "123456789012345678901234" // 24 digits
+	)
+
+	var (
+		content = make([]byte, 0, traceIDLength)
+		epoch   = startTime.Unix()
+	)
+
+	content = append(content, "1-"...)
+	content = strconv.AppendInt(content, epoch, 16)
+	content = append(content, "-"...)
+
+	content = append(content, digits24...)                // expand capacity
+	hex.Encode(content[identifierOffset:], traceID[4:16]) // overwrite with identifier
+
+	return string(content)
+}
+
+// ParseAmazonTraceID parses an amazon traceID string in the format 1-5759e988-bd862e3fe1be46a994272793
+func ParseAmazonTraceID(v string) (trace.TraceID, bool) {
+	if len(v) != traceIDLength {
+		return trace.TraceID{}, false
+	}
+
+	epoch, err := hex.DecodeString(v[epochOffset : epochOffset+8])
+	if err != nil {
+		return trace.TraceID{}, false
+	}
+
+	identifier, err := hex.DecodeString(v[identifierOffset:])
+	if err != nil {
+		return trace.TraceID{}, false
+	}
+
+	var traceID trace.TraceID
+	binary.BigEndian.PutUint32(traceID[0:4], binary.BigEndian.Uint32(epoch))
+	for index, b := range identifier {
+		traceID[index+4] = b
+	}
+
+	return traceID, true
+}
+
+// ParseAmazonSpanID parses an amazon spanID
+func ParseAmazonSpanID(v string) (trace.SpanID, bool) {
+	if len(v) != spanIDLength {
+		return trace.SpanID{}, false
+	}
+
+	data, err := hex.DecodeString(v)
+	if err != nil {
+		return trace.SpanID{}, false
+	}
+
+	var spanID trace.SpanID
+	for index, b := range data {
+		spanID[index] = b
+	}
+
+	return spanID, true
+}
+
+// merge all string, bool, and numeric values from src to dest
+func merge(dest, src map[string]interface{}) {
+	for key, value := range src {
+		switch value.(type) {
+		case bool:
+			dest[key] = value
+		case string:
+			dest[key] = value
+		case int, int8, int16, int32, int64:
+			dest[key] = value
+		case uint, uint8, uint16, uint32, uint64:
+			dest[key] = value
+		case float32, float64:
+			dest[key] = value
+		}
+	}
+}
+
+func makeAnnotations(annotations []trace.Annotation, attributes map[string]interface{}) map[string]interface{} {
+	var result = map[string]interface{}{}
+
+	for _, annotation := range annotations {
```

Comment:
Does AWS not support the notion of attributes/events/logs on a Span?

Reply:
The primary attributes of an xray span are annotations (searchable key values), metadata (non-searchable key values), and exceptions.  Currently, logs and events cannot be associated with spans directly.  We manually insert the traceID into the logs to correlate them.